src/layout.jl

```diff
@@ -635,36 +630,67 @@ PlaceHolder(ren::AbstractRenderable) = PlaceHolder(ren.measure.w, ren.measure.h)
 #                                     GRID                                     #
 # ---------------------------------------------------------------------------- #
 
+"""
+    calc_nrows_ncols(n, aspect=(16, 9))
+
+Computes the number of rows and columns to fit a given number `n` of subplots in a figure with aspect `aspect`.
+Adapted from: stackoverflow.com/a/43366784
+"""
+function calc_nrows_ncols(n, aspect=(16, 9))
+    w, h = aspect
+    factor = √(n / (w * h * 1.0))
+    cols = floor(Int, w * factor)
+    rows = floor(Int, h * factor)
+    rowFirst = w < h
+    while rows * cols < n
+        if rowFirst
+            rows += 1
+        else
+            cols += 1
+        end
+        rowFirst = !rowFirst
+    end
+    return rows, cols
+end
+
 """
     grid(rens::AbstractVector{<:AbstractRenderable}, layout::Union{Nothing,Tuple} = nothing, placeholder = nothing)
 
-Construct a square grid from a `AbsractVector` of `AbstractRenderable`
+Construct a square grid from a `AbsractVector` of `AbstractRenderable`.
 """
 function grid(
     rens::AbstractVector{<:AbstractRenderable};
     layout::Union{Nothing,Tuple} = nothing, 
-    placeholder = nothing
+    placeholder = nothing,
+    aspect = (16, 9),
+    hspacer = "",
+    vspacer = hspacer,
 )
     num = length(rens)
-    if layout === nothing
-        m = ceil(Int, √(num))
+    m, n = layout === nothing ? calc_nrows_ncols(num, aspect) : layout
+    if m === nothing
+        m, r = divrem(num, n)
+        r == 0 || (m += 1)
+    elseif n === nothing
         n, r = divrem(num, m)
         r == 0 || (n += 1)
-    else
-        m, n = layout
     end
-    fill_in = placeholder === nothing ? eltype(rens)() : placeholder
-    return grid(reshape(append!(copy(rens), [fill_in for _ in 1:(m * n - num)]), m, n))
+    fill_in = placeholder === nothing ? PlaceHolder(first(rens)) : placeholder
+    return grid(
+        reshape(append!(copy(rens), [fill_in for _ in 1:(m * n - num)]), m, n);
+        hspacer = hspacer,
+        vspacer = vspacer,
+    )
 end
 
 """
     grid(rens::AbstractMatrix{<:AbstractRenderable})
 
-Construct a grid from a `AbstractMatrix` of `AbstractRenderable`
+Construct a grid from a `AbstractMatrix` of `AbstractRenderable`.
 """
-function grid(rens::AbstractMatrix{<:AbstractRenderable})
-    rows = collect(foldl(*, col[2:end], init = first(col)) for col in eachrow(rens))
-    return foldl(/, rows[2:end], init = first(rows))
+function grid(rens::AbstractMatrix{<:AbstractRenderable}; hspacer = "", vspacer = hspacer)
```

Comment:
Unlike Python, the default values in Julia are set when called instead of when defined:

```julia
julia> f(x, y=x) = y
f (generic function with 2 methods)

julia> f(1)
1

julia> f(2)
2
```

Reply:
Sorry if I can suggest one additional change, I think vertical/horizontal spacing (or padding) should be expressed as integers: the amount of space required. It would make more sense for the user, then you can create the strings under the hood.

Reply:
Hey @johnnychen94  nice to see you here! I understand now, beautiful - thanks

Reply:
> you can add a previous keyword argument as the value for another kehword argument

It's a common thing in julia to apply a function on given keyword to compute a default value e.g. https://github.com/JuliaPlots/Plots.jl/blob/b959e64b9c189c1e5383582f7d1963a27e76c975/src/axes.jl#L594.

> Can they be set independently

yes.

> instead of h/vspacer

Renamed to `hpad`, `vpad`.

> I think vertical/horizontal spacing (or padding) should be expressed as integers:

Will fix.

Reply:
I had no idea! I can see how that could be very useful, nice to know!

And thanks for making the changes